core/trino-main/src/main/java/io/trino/operator/output/BlockTypeDispatchingPositionsAppender.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Dispatches the {@link #append} and {@link #appendRle} methods to the {@link #delegate} depending on the input {@link Block} class.
+ * The {@link Block} is flattened if necessary so that the {@link #delegate} {@link PositionsAppender#append(IntArrayList, Block)}
+ * always gets flat {@link Block} and {@link PositionsAppender#appendRle(RunLengthEncodedBlock)} always gets {@link RunLengthEncodedBlock}
+ * with {@link RunLengthEncodedBlock#getValue()} being flat {@link Block}.
+ */
+public class BlockTypeDispatchingPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockTypeDispatchingPositionsAppender.class).instanceSize();
+
+    private final PositionsAppender delegate;
+
+    public BlockTypeDispatchingPositionsAppender(PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+
+        if (source instanceof RunLengthEncodedBlock) {
```

Comment:
actually, generic `append` should not be called with `RunLengthEncodedBlock`, so you can just do check state here

Reply:
this is exactly what this method should do - dispatch based on the block type (including DictionaryBlock).
it doesn't make sense for every client of this class to do the dispatching.

Reply:
> this is exactly what this method should do - dispatch based on the block type (including DictionaryBlock).

This makes the contract of `PositionsAppender` ambiguous. There is `appendRle` which accepts RLE, but also `append` which also can accept RLE. In fact, I would maybe rename `append` to `appendFlat`

> it doesn't make sense for every client of this class to do the dispatching.

there is just one client